src/peft/tuners/adamix.py

```diff
@@ -0,0 +1,416 @@
+        for key in key_list:
+            try:
+                parent, target, target_name = _get_submodules(self.model, key)
+            except AttributeError:
+                continue
+            if isinstance(target, ExpertSoup):
+                delattr(parent, target_name)
+        return self.model
+
+    def get_two_view_from_model(self) -> Tuple[Tuple]:
+        # NOTE: To perform backward pass on the two views, you must set retain_graph=True in the optimizer
+
+        two_views = []
+        for name, module in self.model.named_modules():
+            if isinstance(module, ExpertSoup):
+                two_views.append(module.two_views)
+
+        if len(two_views) == 0:
+            raise ValueError("Returned views are empty")
+
+        return two_views
+
+
+def mark_only_adamix_as_trainable(model: nn.Module) -> None:
+    for n, p in model.named_parameters():
+        if "adamix" not in n:
+            p.requires_grad = False
+
+
+# # Below code is based on https://github.com/microsoft/AdaMix
+
+
+# #  ------------------------------------------------------------------------------------------
+# #  Copyright (c) Microsoft Corporation. All rights reserved.
+# #  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+# #  ------------------------------------------------------------------------------------------
+
+
+class MixtureSoup(nn.Module):
+    def __init__(self, expert, adapter_name, inference_mode=False, num_local_experts=1):
+        super().__init__()
+        self.inference_mode = inference_mode
+        self.dict_keys = ["_".join([adapter_name, str(i)]) for i in range(num_local_experts)]
+        self.expert_mixture = nn.ModuleDict({})
+        for i in self.dict_keys:
+            self.expert_mixture[i] = copy.deepcopy(expert)
+        self.num_local_experts = num_local_experts
+
+    def get_expert_by_idx(self, idx):
+        return self.expert_mixture[self.dict_keys[idx]]
+
+    def expert_soup_forward(self, x):
+        output = F.linear(x, self.parameter_dict["weight"], self.parameter_dict["bias"])
+        return output
+
+    def expert_soup(self):
+        self.parameter_dict = {"weight": 0, "bias": 0}
+        for idx in range(self.num_local_experts):
+            single_expert = self.expert_mixture[idx]
+            for s_name, s_param in single_expert.named_parameters():
+                if "weight" in s_name:
+                    p_name = "weight"
+                else:
+                    p_name = "bias"
+                self.parameter_dict[p_name] = self.parameter_dict[p_name] + s_param / self.num_local_experts
+
+    def forward(self, x: torch.Tensor):
+        expert_output = None
+
+        if not self.inference_mode:
+            expert_idx = torch.randint(low=0, high=self.num_local_experts, size=(1,)).item()  # selected expert
+            expert_output = self.get_expert_by_idx(expert_idx)(x)
+
+        else:
```

Comment:
According to test coverage, these lines are never tested, i.e. there are no tests for `inference_mode`. Could you please add some?

Reply:
Changed this to be triggered when we do model.eval()